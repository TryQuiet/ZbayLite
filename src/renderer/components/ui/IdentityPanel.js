import React from 'react'
import PropTypes from 'prop-types'
import * as R from 'ramda'

import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'
import { Button } from '@material-ui/core'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'

import SettingsModal from '../../containers/widgets/settings/SettingsModal'
import ReceivedInvitationModal from '../../containers/ui/InvitationModal/ReceivedInvitationModal'
import CreateUsernameModal from '../../containers/widgets/createUsernameModal/CreateUsername'
// import ImportChannelModal from '../../containers/widgets/channels/importChannel/ImportChannelModal'

const styles = theme => ({
  root: {
    marginTop: theme.spacing(1),
    WebkitAppRegion: process.platform === 'win32' ? 'no-drag' : 'drag',
    paddingLeft: 16,
    paddingRight: 16
  },

  button: {
    color: theme.palette.colors.white,
    padding: 0,
    textAlign: 'left',
    opacity: 0.8,
    '&:hover': {
      opacity: 1,
      backgroundColor: 'inherit'
    }
  },
  buttonLabel: {
    justifyContent: 'flex-start',
    textTransform: 'none'
  },
  nickname: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    maxWidth: 175,
    whiteSpace: 'nowrap'
  }
})

export const IdentityPanel = ({ classes, identity, handleSettings, user }) => {
  const nickname = user ? user.nickname : `anon${identity.signerPubKey.substring(0, 10)}`
  return (
    <div className={classes.root}>
      <Button
        onClick={handleSettings}
        component='span'
        classes={{ root: classes.button, label: classes.buttonLabel }}
      >
        <Typography variant='h4' className={classes.nickname}>{nickname}</Typography>
        <ExpandMoreIcon fontSize='small' />
      </Button>
      <SettingsModal />
      <ReceivedInvitationModal />
      {/* <ImportChannelModal /> */}
      <CreateUsernameModal />
    </div>
  )
}

IdentityPanel.propTypes = {
  classes: PropTypes.object.isRequired,
  identity: PropTypes.shape({
    name: PropTypes.string.isRequired,
    address: PropTypes.string.isRequired,
    signerPubKey: PropTypes.string.isRequired
  }).isRequired,
  handleSettings: PropTypes.func.isRequired,
  user: PropTypes.shape({
    nickname: PropTypes.string.isRequired
  }).isRequired
}

export default R.compose(
  React.memo,
  withStyles(styles)
)(IdentityPanel)
