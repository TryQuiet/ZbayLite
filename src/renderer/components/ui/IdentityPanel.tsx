import React from 'react'

import Typography from '@material-ui/core/Typography'
import { Button } from '@material-ui/core'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'

import SettingsModal from '../../containers/widgets/settings/SettingsModal'
import ReceivedInvitationModal from '../../containers/ui/InvitationModal/ReceivedInvitationModal'
import CreateUsernameModal from '../../containers/widgets/createUsernameModal/CreateUsername'

import { makeStyles, MuiThemeProvider } from '@material-ui/core/styles'

import theme from '../../theme'

const useStyles = makeStyles({
  root: {
    marginTop: theme.spacing(1),
    WebkitAppRegion: process.platform === 'win32' ? 'no-drag' : 'drag',
    paddingLeft: 16,
    paddingRight: 16
  },
  button: {
    color: theme.palette.colors.white,
    padding: 0,
    textAlign: 'left',
    opacity: 0.8,
    '&:hover': {
      opacity: 1,
      backgroundColor: 'inherit'
    }
  },
  buttonLabel: {
    justifyContent: 'flex-start',
    textTransform: 'none'
  },
  nickname: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    maxWidth: 175,
    whiteSpace: 'nowrap'
  }
})

interface IdentityPanelProps {
  user: User
  identity: Identity
  handleSettings: () => void
}

interface Identity {
  name: string
  address: string
  signerPubKey: string
}

interface User {
  nickname: string
}

export const IdentityPanel: React.FC<IdentityPanelProps> = ({ identity, handleSettings, user }) => {
  const classes = useStyles({})

  const nickname = user ? user.nickname : `anon${identity.signerPubKey.substring(0, 10)}`

  return (
    <MuiThemeProvider theme={theme}>
      <div className={classes.root}>
        <Button
          onClick={handleSettings}
          component='span'
          classes={{ root: classes.button, label: classes.buttonLabel }}>
          <Typography variant='h4' className={classes.nickname}>
            {nickname}
          </Typography>
          <ExpandMoreIcon fontSize='small' />
        </Button>
        <SettingsModal />
        <ReceivedInvitationModal />
        {/* <ImportChannelModal /> */}
        <CreateUsernameModal />
      </div>
    </MuiThemeProvider>
  )
}

export default IdentityPanel
