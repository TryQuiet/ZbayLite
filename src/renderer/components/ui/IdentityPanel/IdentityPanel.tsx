import React from 'react'

import Typography from '@material-ui/core/Typography'
import { Button } from '@material-ui/core'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'

import SettingsModal from '../../../containers/widgets/settings/SettingsModal'
import ReceivedInvitationModal from '../../../containers/ui/InvitationModal/ReceivedInvitationModal'
import CreateUsernameModal from '../../../containers/widgets/createUsernameModal/CreateUsername'

import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles(theme => ({
  root: {
    marginTop: theme.spacing(1),
    WebkitAppRegion: process.platform === 'win32' ? 'no-drag' : 'drag',
    paddingLeft: 16,
    paddingRight: 16
  },
  button: {
    color: theme.palette.colors.white,
    padding: 0,
    textAlign: 'left',
    opacity: 0.8,
    '&:hover': {
      opacity: 1,
      backgroundColor: 'inherit'
    }
  },
  buttonLabel: {
    justifyContent: 'flex-start',
    textTransform: 'none'
  },
  nickname: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    maxWidth: 175,
    whiteSpace: 'nowrap'
  }
}))

interface IdentityPanelProps {
  user: User
  identity: Identity
  handleSettings: () => void
}

interface Identity {
  name: string
  address: string
  signerPubKey: string
}

interface User {
  nickname: string
}

export const IdentityPanel: React.FC<IdentityPanelProps> = ({ identity, handleSettings, user }) => {
  const classes = useStyles({})

  const nickname = user ? user.nickname : `anon${identity.signerPubKey.substring(0, 10)}`

  return (
    <div className={classes.root}>
      <Button
        onClick={handleSettings}
        component='span'
        classes={{ root: classes.button, label: classes.buttonLabel }}>
        <Typography variant='h4' className={classes.nickname}>
          {nickname}
        </Typography>
        <ExpandMoreIcon fontSize='small' />
      </Button>
      <SettingsModal />
      <ReceivedInvitationModal />
      {/* <ImportChannelModal /> */}
      <CreateUsernameModal />
    </div>
  )
}

export default IdentityPanel
