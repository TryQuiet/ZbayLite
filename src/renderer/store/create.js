import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
