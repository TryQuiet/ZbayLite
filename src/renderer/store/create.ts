import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
