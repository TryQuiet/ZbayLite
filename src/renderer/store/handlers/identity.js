import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'
import secp256k1 from 'secp256k1'
import { randomBytes } from 'crypto'
import { ipcRenderer } from 'electron'
import { DateTime } from 'luxon'

import client from '../../zcash'
import channels from '../../zcash/channels'

import identitySelectors from '../selectors/identity'
// import nodeSelectors from '../selectors/node'
import txnTimestampsSelector from '../selectors/txnTimestamps'
// import channelsSelectors from '../selectors/channels'
// import channelsHandlers from './channels'
// import removedChannelsHandlers from './removedChannels'
// import usersHandlers from './users'
// import contactsHandlers from './contacts'
// import messagesHandlers from './messages'
// import publicChannelsHandlers from './publicChannels'
import coordinatorHandlers from './coordinator'
// import offersHandlers from './offers'
import whitelistHandlers from './whitelist'
import ownedChannelsHandlers from './ownedChannels'
import txnTimestampsHandlers from './txnTimestamps'
import logsHandlers from '../../store/handlers/logs'
// import vaultHandlers from './vault'
import ratesHandlers from './rates'
import nodeHandlers from './node'
import notificationCenterHandlers from './notificationCenter'
// import migrateTo_0_2_0 from '../../../shared/migrations/0_2_0' // eslint-disable-line camelcase
// import migrateTo_0_7_0 from '../../../shared/migrations/0_7_0' // eslint-disable-line camelcase
import { LoaderState, successNotification } from './utils'
import modalsHandlers from './modals'
import notificationsHandlers from './notifications'
import messagesHandlers from './messages'
import {
  actionTypes,
  networkFeeSatoshi,
  satoshiMultiplier
} from '../../../shared/static'
import electronStore from '../../../shared/electronStore'
// import channels from '../../zcash/channels'

export const ShippingData = Immutable.Record(
  {
    firstName: '',
    lastName: '',
    street: '',
    country: '',
    region: '',
    city: '',
    postalCode: ''
  },
  'ShippingData'
)

const RemovedChannels = Immutable.Record(
  {
    removedChannels: Immutable.List()
  },
  'RemovedChannels'
)

const Identity = Immutable.Record(
  {
    id: null,
    address: '',
    transparentAddress: '',
    signerPrivKey: '',
    signerPubKey: '',
    name: '',
    shippingData: ShippingData(),
    balance: null,
    lockedBalance: null,
    donationAllow: true,
    shieldingTax: true,
    donationAddress: '',
    freeUtxos: 0,
    addresses: Immutable.List(),
    shieldedAddresses: Immutable.List()
  },
  'Identity'
)

// export const Identity = values => {
//   const record = _Identity(values)
//   return record.set('shippingData', ShippingData(record.shippingData))
// }

export const IdentityState = Immutable.Record(
  {
    data: Identity(),
    fetchingBalance: false,
    loader: LoaderState({ loading: false }),
    removedChannels: RemovedChannels(),
    errors: ''
  },
  'IdentityState'
)

export const initialState = IdentityState()

export const setIdentity = createAction(actionTypes.SET_IDENTITY)
export const setRemovedChannels = createAction(actionTypes.SET_REMOVED_CHANNELS)
export const setBalance = createAction(actionTypes.SET_IDENTITY_BALANCE)
export const setLockedBalance = createAction(
  actionTypes.SET_IDENTITY_LOCKED_BALANCE
)
export const setErrors = createAction(actionTypes.SET_IDENTITY_ERROR)
export const setFetchingBalance = createAction(actionTypes.SET_FETCHING_BALANCE)
export const setLoading = createAction(actionTypes.SET_IDENTITY_LOADING)
export const setLoadingMessage = createAction(
  actionTypes.SET_IDENTITY_LOADING_MESSAGE
)
export const setShippingData = createAction(
  actionTypes.SET_IDENTITY_SHIPPING_DATA
)
export const setDonationAllow = createAction(actionTypes.SET_DONATION_ALLOW)
export const setDonationAddress = createAction(actionTypes.SET_DONATION_ADDRESS)
export const setShieldingTax = createAction(actionTypes.SET_SHIELDING_TAX)
export const setFreeUtxos = createAction(actionTypes.SET_FREE_UTXOS)
export const setUserAddreses = createAction(actionTypes.SET_USER_ADDRESSES)
export const setUserShieldedAddreses = createAction(
  actionTypes.SET_USER_SHIELDED_ADDRESES
)

const actions = {
  setIdentity,
  setRemovedChannels,
  setFetchingBalance,
  setErrors,
  setLoading,
  setLoadingMessage,
  setLockedBalance,
  setShippingData,
  setBalance,
  setDonationAllow,
  setDonationAddress,
  setShieldingTax,
  setFreeUtxos
}
export const fetchAffiliateMoney = () => async (dispatch, getState) => {
  try {
    const identityAddress = identitySelectors.address(getState())
    const transfers = await client().payment.received(identityAddress)
    const affiliatesTransfers = transfers.filter(msg =>
      msg.memo.startsWith('aa')
    )
    let amount = 0
    let txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    for (const key in affiliatesTransfers) {
      const transfer = transfers[key]
      if (!txnTimestamps.get(transfer.txid)) {
        amount += transfer.amount
        const result = await client().confirmations.getResult(transfer.txid)
        await dispatch(
          txnTimestampsHandlers.actions.addTxnTimestamp({
            tnxs: { [transfer.txid]: result.timereceived.toString() }
          })
        )
      }
    }
    if (amount) {
      dispatch(
        notificationsHandlers.actions.enqueueSnackbar(
          successNotification({
            message: `You received ${amount} ZEC from your affiliates`
          })
        )
      )
    }
  } catch (err) {}
}
export const fetchBalance = () => async (dispatch, getState) => {
  try {
    dispatch(setFetchingBalance(true))
    const balanceObj = await client.balance()
    const notes = await client.notes()
    const pending = notes.pending_notes.reduce((acc, cur) => acc + cur.value, 0)
    dispatch(
      setLockedBalance(
        new BigNumber(
          (balanceObj.unverified_zbalance + pending) / satoshiMultiplier
        )
      )
    )
    dispatch(
      setBalance(
        new BigNumber(balanceObj.spendable_zbalance / satoshiMultiplier)
      )
    )
    dispatch(
      logsHandlers.epics.saveLogs({
        type: 'APPLICATION_LOGS',
        payload: `Fetching balance: locked balance: ${(balanceObj.unverified_zbalance +
          pending) /
          satoshiMultiplier}, balance: ${balanceObj.spendable_zbalance /
          satoshiMultiplier}`
      })
    )
  } catch (err) {
    dispatch(setErrors(err.message))
  } finally {
    dispatch(setFetchingBalance(false))
  }
}
export const fetchFreeUtxos = () => async (dispatch, getState) => {
  try {
    const utxos = await client.notes()
    const freeUtxos = utxos.unspent_notes.filter(
      utxo => utxo.value > networkFeeSatoshi && utxo.spendable === true
    )
    dispatch(setFreeUtxos(freeUtxos.length))
    dispatch(
      logsHandlers.epics.saveLogs({
        type: 'APPLICATION_LOGS',
        payload: `Setting free UTXOs: ${freeUtxos.length}`
      })
    )
  } catch (err) {
    console.warn(err)
  }
}

export const createSignerKeys = () => {
  let signerPrivKey
  do {
    signerPrivKey = randomBytes(32)
  } while (!secp256k1.privateKeyVerify(signerPrivKey))
  return {
    signerPrivKey: signerPrivKey.toString('hex'),
    signerPubKey: secp256k1.publicKeyCreate(signerPrivKey, true).toString('hex')
  }
}

export const createIdentity = ({ name }) => async (dispatch, getState) => {
  try {
    const accountAddresses = await client.addresses()
    const { z_addresses: zAddresses } = accountAddresses
    const { t_addresses: tAddresses } = accountAddresses
    const [zAddress] = zAddresses
    const [tAddress] = tAddresses
    const tpk = await client.getPrivKey(tAddress)
    const sk = await client.getPrivKey(zAddress)

    const { signerPrivKey, signerPubKey } = exportFunctions.createSignerKeys()

    electronStore.set('identity', {
      name,
      address: zAddress,
      transparentAddress: tAddress,
      signerPubKey,
      signerPrivKey,
      keys: {
        tpk,
        sk
      }
    })
    const network = 'mainnet'

    const channelsToImport = [
      'general',
      'registeredUsers',
      'channelOfChannels',
      'store',
      'priceOracle'
    ]
    const channelsWithDetails = channelsToImport.reduce((o, key) => {
      const channelDetails = channels[key][network]
      const preparedChannel = {
        name: channelDetails.name,
        private: channelDetails.private,
        address: channelDetails.address,
        unread: channelDetails.unread,
        description: channelDetails.description,
        keys: channelDetails.keys,
        lastSeen: DateTime.utc().toSeconds()
      }
      return { ...o, [preparedChannel.address]: preparedChannel }
    }, {})
    electronStore.set('defaultChannels', channelsWithDetails)
    dispatch(
      logsHandlers.epics.saveLogs({
        type: 'APPLICATION_LOGS',
        payload: `Creating identity / importing default channels`
      })
    )
    const channelsToLoad = Object.keys(channelsWithDetails)
    for (const channel of channelsToLoad) {
      await client.importKey(
        channelsWithDetails[channel]['keys']['ivk'],
        740000
      )
    }
    await client.rescan()
    return electronStore.get('identity')
  } catch (err) {
    console.log('error', err)
  }
}

export const loadIdentity = () => async (dispatch, getState) => {
  const identity = electronStore.get('identity')
  if (identity) {
    await dispatch(setIdentity(identity))
    dispatch(
      logsHandlers.epics.saveLogs({
        type: 'APPLICATION_LOGS',
        payload: `Loading identity`
      })
    )
  }
}

export const createWalletBackup = () => async (dispatch, getState) => {
  const isDev = process.env.NODE_ENV === 'development'
  const isWalletCopyCreated = electronStore.get('isWalletCopyCreated')
  if (!isWalletCopyCreated && !isDev) {
    ipcRenderer.send('make-wallet-backup')
  }
}

export const setIdentityEpic = (identityToSet, isNewUser) => async (
  dispatch,
  getState
) => {
  // let identity = await migrateTo_0_2_0.ensureIdentityHasKeys(identityToSet)
  let identity = identityToSet
  dispatch(setLoading(true))
  dispatch(
    logsHandlers.epics.saveLogs({
      type: 'APPLICATION_LOGS',
      payload: `Start loading identity`
    })
  )
  // const isRescanned = electronStore.get('AppStatus.blockchain.isRescanned')
  const isNewUser = electronStore.get('isNewUser')
  try {
    const removedChannels = electronStore.get('removedChannels')
    if (removedChannels) {
      const removedChannelsList = Immutable.fromJS(Object.keys(removedChannels))
      dispatch(setRemovedChannels(removedChannelsList))
    }
    dispatch(setLoadingMessage('Ensuring identity integrity'))
    await dispatch(setLoadingMessage('Ensuring node contains identity keys'))
    await dispatch(whitelistHandlers.epics.initWhitelist())
    await dispatch(notificationCenterHandlers.epics.init())
    dispatch(setLoadingMessage('Setting identity'))
    await dispatch(setIdentity(identity))
    const shippingAddress = electronStore.get('identity.shippingData')
    if (shippingAddress) {
      dispatch(setShippingData(ShippingData(shippingAddress)))
    }
    dispatch(setLoadingMessage('Fetching balance and loading channels'))
    await dispatch(initAddreses())
    dispatch(ownedChannelsHandlers.epics.getOwnedChannels())
    dispatch(ratesHandlers.epics.setInitialPrice())
    await dispatch(nodeHandlers.epics.getStatus())
    await dispatch(fetchBalance())
    await dispatch(fetchFreeUtxos())
    await dispatch(messagesHandlers.epics.fetchMessages())
    setTimeout(() => dispatch(coordinatorHandlers.epics.coordinator()), 5000)

    dispatch(setLoadingMessage('Loading users and messages'))
  } catch (err) {}
  const zecBalance = identitySelectors.balance('zec')(getState())
  if (isNewUser === true && zecBalance.gt(0)) {
    dispatch(modalsHandlers.actionCreators.openModal('createUsernameModal')())
  }
  // dispatch(fetchAffiliateMoney())
  dispatch(setLoading(false))
  dispatch(
    logsHandlers.epics.saveLogs({
      type: 'APPLICATION_LOGS',
      payload: ` Loading identity finished`
    })
  )
  // Don't show deposit modal if we use faucet 12.02.2020
  // const balance = identitySelectors.balance('zec')(getState())
  // const lockedBalance = identitySelectors.lockedBalance('zec')(getState())
  // if (lockedBalance.plus(balance).lt(0.0001) && newUser === false) {
  //   setTimeout(
  //     () => dispatch(modalsHandlers.actionCreators.openModal('depositMoney')()),
  //     500
  //   )
  // }
}

export const updateShippingData = (values, formActions) => async (
  dispatch,
  getState
) => {
  electronStore.set('identity.shippingData', values)
  await dispatch(setShippingData(ShippingData(values)))
  dispatch(
    notificationsHandlers.actions.enqueueSnackbar(
      successNotification({ message: 'Shipping Address Updated' })
    )
  )
  dispatch(
    logsHandlers.epics.saveLogs({
      type: 'APPLICATION_LOGS',
      payload: `Updating shipping data`
    })
  )
  formActions.setSubmitting(false)
}

export const updateDonation = allow => async (dispatch, getState) => {
  dispatch(
    notificationsHandlers.actions.enqueueSnackbar(
      successNotification({ message: 'Donation information updated' })
    )
  )
  dispatch(
    logsHandlers.epics.saveLogs({
      type: 'APPLICATION_LOGS',
      payload: `Updating donation status`
    })
  )
}

export const updateDonationAddress = address => async (dispatch, getState) => {}
export const updateShieldingTax = allow => async (dispatch, getState) => {}
export const generateNewAddress = () => async (dispatch, getState) => {
  if (!electronStore.get('addresses')) {
    electronStore.set('addresses', JSON.stringify([]))
  }
  const addresses = JSON.parse(electronStore.get('addresses'))
  const address = await client().addresses.createTransparent()
  addresses.unshift(address)
  dispatch(setUserAddreses(Immutable.List(addresses)))
  electronStore.set('addresses', JSON.stringify(addresses))
}
export const generateNewShieldedAddress = () => async (dispatch, getState) => {
  if (!electronStore.get('shieldedAddresses')) {
    electronStore.set('shieldedAddresses', JSON.stringify([]))
  }
  const addresses = JSON.parse(electronStore.get('shieldedAddresses'))
  const address = await client().addresses.create('sapling')
  addresses.unshift(address)
  dispatch(setUserShieldedAddreses(Immutable.List(addresses)))
  electronStore.set('shieldedAddresses', JSON.stringify(addresses))
}
export const initAddreses = () => async (dispatch, getState) => {
  if (!electronStore.get('addresses')) {
    electronStore.set('addresses', JSON.stringify([]))
  }
  if (!electronStore.get('shieldedAddresses')) {
    electronStore.set('shieldedAddresses', JSON.stringify([]))
  }
  const addresses = JSON.parse(electronStore.get('addresses'))
  dispatch(setUserAddreses(Immutable.List(addresses)))
  const shieldedAddreses = JSON.parse(electronStore.get('shieldedAddresses'))
  dispatch(setUserShieldedAddreses(Immutable.List(shieldedAddreses)))
}
const epics = {
  fetchBalance,
  createIdentity,
  setIdentity: setIdentityEpic,
  updateShippingData,
  createSignerKeys,
  updateDonation,
  updateDonationAddress,
  updateShieldingTax,
  fetchFreeUtxos,
  loadIdentity,
  generateNewAddress,
  generateNewShieldedAddress
}

const exportFunctions = {
  createSignerKeys
}

export const reducer = handleActions(
  {
    [setLoading]: (state, { payload: loading }) =>
      state.setIn(['loader', 'loading'], loading),
    [setRemovedChannels]: (state, { payload: removedChannels }) =>
      state.set('removedChannels', removedChannels),
    [setLoadingMessage]: (state, { payload: message }) =>
      state.setIn(['loader', 'message'], message),
    [setIdentity]: (state, { payload: identity }) =>
      state.update('data', data => data.merge(identity)),
    [setBalance]: (state, { payload: balance }) =>
      state.update('data', data => data.set('balance', balance)),
    [setLockedBalance]: (state, { payload: balance }) =>
      state.update('data', data => data.set('lockedBalance', balance)),
    [setFetchingBalance]: (state, { payload: fetching }) =>
      state.set('fetchingBalance', fetching),
    [setErrors]: (state, { payload: errors }) => state.set('errors', errors),
    [setShippingData]: (state, { payload: shippingData }) =>
      state.setIn(['data', 'shippingData'], ShippingData(shippingData)),
    [setDonationAllow]: (state, { payload: allow }) =>
      state.setIn(['data', 'donationAllow'], allow),
    [setDonationAddress]: (state, { payload: address }) =>
      state.setIn(['data', 'donationAddress'], address),
    [setShieldingTax]: (state, { payload: allow }) =>
      state.setIn(['data', 'shieldingTax'], allow),
    [setFreeUtxos]: (state, { payload: freeUtxos }) =>
      state.setIn(['data', 'freeUtxos'], freeUtxos),
    [setUserAddreses]: (state, { payload: addresses }) =>
      state.setIn(['data', 'addresses'], addresses),
    [setUserShieldedAddreses]: (state, { payload: shieldedAddresses }) =>
      state.setIn(['data', 'shieldedAddresses'], shieldedAddresses)
  },
  initialState
)

export default {
  actions,
  epics,
  reducer,
  exportFunctions
}
