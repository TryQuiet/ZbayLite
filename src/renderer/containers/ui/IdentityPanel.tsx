import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import * as R from 'ramda'

import IdentityPanel from '../../components/ui/IdentityPanel'
import identitySelectors from '../../store/selectors/identity'
import usersSelectors from '../../store/selectors/users'
import { actionCreators } from '../../store/handlers/modals'

export const mapStateToProps = state => {
  const identity = identitySelectors.data(state)
  return {
    identity: identity,
    user: usersSelectors.registeredUser(identity.signerPubKey)(state)
  }
}

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      handleSettings: actionCreators.openModal('accountSettingsModal'),
      handleInvitation: actionCreators.openModal('invitationModal')
    },
    dispatch
  )

export default R.compose(
  React.memo,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(IdentityPanel)
