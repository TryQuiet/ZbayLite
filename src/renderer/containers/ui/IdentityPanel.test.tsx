import { useData } from './IdentityPanel'

import React from 'react'
import { renderHook } from '@testing-library/react-hooks'
import { Provider } from 'react-redux'

import create from '../../store/create'
import { initialState } from '../../store/handlers/identity'

describe('IdentityPanel', () => {
  let store = null
  let wrapper
  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      identity: {
        ...initialState,
        data: {
          ...initialState.data,
          address: 'zctestaddress',
          balance: '23.435432',
          lockedBalance: '13.123432',
          name: 'saturn'
        }
      }
    })
    wrapper = ({ children }) => <Provider store={store}>{children}</Provider>
  })

  it('will receive right props', async () => {
    const { result } = renderHook(() => useData(), { wrapper })
    expect(result.current).toMatchSnapshot()
  })
})
