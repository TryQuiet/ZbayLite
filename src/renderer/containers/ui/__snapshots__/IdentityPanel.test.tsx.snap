// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentityPanel will receive right props 1`] = `
Object {
  "identity": undefined,
}
`;
